### DISCUSSION TITLE
Running METplus on Orion

### DISCUSSION THREAD

**Discussion by briannen** [dtcenter/METplus#2129] [Incoming] [answered]
I am currently trying to run METplus 5.0.1 and MET 11.0.1 on Orion using the SRW workflow. METplus is finding all the data correctly, but is failing when it is trying to actually run the MET tool. I'm getting the following error in my log file:

/apps/contrib/MET/11.0.1/bin/grid_stat: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

I also attempted to run with METplus 5.0.0 and MET 11.0.0, which was working previously, but it is failing with the same error.

Any ideas?

**Comment by jprestop** [comment]
HI @briannen.  Thank you for letting us know about this problem.  Looking at the [Existing Builds](https://dtcenter.org/community-code/metplus/metplus-5-0-existing-builds) page, I am using the following to load METplus on Orion:
```
module use /apps/contrib/modulefiles
module load metplus/5.0.1
```
When I run grid_stat, I do not get an error:
```
(base) Orion-devel-1[3] role-ovp$ grid_stat
DEBUG 1: Start grid_stat by role-ovp(9175) at 2023-04-19 18:47:07Z  cmd: grid_stat 
DEBUG 2: OMP_NUM_THREADS is not set. Defaulting to 1 thread. Recommend setting OMP_NUM_THREADS for faster runtimes.
DEBUG 2: OpenMP running on 1 thread(s).

*** Model Evaluation Tools (METV11.0.1) ***

Usage: grid_stat
	fcst_file
	obs_file
	config_file
	[-outdir path]
	[-log file]
	[-v level]
	[-compress level]

	where	"fcst_file" is a gridded forecast file containing the field(s) to be verified (required).
		"obs_file" is a gridded observation file containing the verifying field(s) (required).
		"config_file" is a GridStatConfig file containing the desired configuration settings (required).
		"-outdir path" overrides the default output directory (.) (optional).
		"-log file" outputs log messages to the specified file (optional).
		"-v level" overrides the default level of logging (2) (optional).
		"-compress level" overrides the compression level of NetCDF variable (0) (optional).
```
Can you please tell me more information about your environment and how you are loading METplus?

**Reply by briannen** [reply]
I'm running through the SRW app workflow, so I don't actually load METplus directly.

**Reply by jprestop** [reply]
@briannen I am not familiar with the SRW app workflow.  Is there a way to determine how METplus is loaded in the workflow?  Could you provide more information about where the SRW files are located?

**Reply by briannen** [reply]
I'm not entirely sure how it is being loaded (I'm not super familiar either). I have since tried to load metplus directly using the modules on the existing builds page, and it's still giving me the same error. Even just trying to run grid stat on it's own gives the same error.

(regional_workflow) Orion-login-2[138] bnelson$ module load contrib
(regional_workflow) Orion-login-2[139] bnelson$ module load intel/2020.2
(regional_workflow) Orion-login-2[140] bnelson$ module load met/11.0.1
(regional_workflow) Orion-login-2[141] bnelson$ /apps/contrib/MET/11.0.1/bin/grid_stat 
/apps/contrib/MET/11.0.1/bin/grid_stat: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

**Reply by jprestop** [reply]
Thank you, @briannen.  I can't seem to reproduce the problem you are experiencing:
```
Orion-devel-1[2] jprestop$ module load contrib
Orion-devel-1[3] jprestop$ module load intel/2020.2
Orion-devel-1[4] jprestop$ module load met/11.0.1
Orion-devel-1[5] jprestop$ grid_stat
DEBUG 1: Start grid_stat by jprestop(8132) at 2023-04-19 19:22:58Z  cmd: grid_stat 
DEBUG 2: OMP_NUM_THREADS is not set. Defaulting to 1 thread. Recommend setting OMP_NUM_THREADS for faster runtimes.
DEBUG 2: OpenMP running on 1 thread(s).

*** Model Evaluation Tools (METV11.0.1) ***

Usage: grid_stat
	fcst_file
	obs_file
	config_file
	[-outdir path]
	[-log file]
	[-v level]
	[-compress level]

	where	"fcst_file" is a gridded forecast file containing the field(s) to be verified (required).
		"obs_file" is a gridded observation file containing the verifying field(s) (required).
		"config_file" is a GridStatConfig file containing the desired configuration settings (required).
		"-outdir path" overrides the default output directory (.) (optional).
		"-log file" outputs log messages to the specified file (optional).
		"-v level" overrides the default level of logging (2) (optional).
		"-compress level" overrides the compression level of NetCDF variable (0) (optional).
```
which leads me to think it is something in your environment.  I'm just not sure what.  Could you please try running these commands outside of the "regional_workflow" environment that you are in currently?

**Reply by briannen** [reply]
I opened up a new window on orion just to make sure that there was nothing extra loaded this time.

(base) Orion-login-1[1] bnelson$ module load contrib
(base) Orion-login-1[2] bnelson$ module load intel/2020.2
(base) Orion-login-1[3] bnelson$ module load met/11.0.1
(base) Orion-login-1[4] bnelson$ /apps/contrib/MET/11.0.1/bin/grid_stat
/apps/contrib/MET/11.0.1/bin/grid_stat: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

**Reply by jprestop** [reply]
@briannen Interesting.  Are you in an existing conda environment (base) when you run these commands?

**Reply by briannen** [reply]
Nope, not in any environment.

**Reply by jprestop** [reply]
@briannen Thank you.  Could you please open a clean shell, run the module commands and then run "module list" and send the result so we can see what is loaded?  Here is my result:
```
Orion-devel-1[6] jprestop$ module list

Currently Loaded Modules:
  1) contrib/0.1   2) intel/2020.2   3) met/11.0.1
```

**Reply by jprestop** [reply]
@briannen Could you also please run "echo $PATH" and send that result?  Here is mine:
```
Orion-devel-1[7] jprestop$ echo $PATH
/apps/contrib/MET/11.0.1/bin:/apps/contrib/MET/11.0.1/external_libs/bin:/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/bin:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin/intel64:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin:/apps/intel-2020.2/intel-2020.2/debugger_2020/gdb/intel64/bin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/lib64/qt-3.3/bin:/apps/sbin:/apps/bin:/opt/slurm/bin
```

**Reply by briannen** [reply]
(base) Orion-login-1[5] bnelson$ module list

Currently Loaded Modules:
  1) contrib/0.1   2) intel/2020.2   3) met/11.0.1

 

(base) Orion-login-1[6] bnelson$ echo $PATH
/apps/contrib/MET/11.0.1/bin:/apps/contrib/MET/11.0.1/external_libs/bin:/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/bin:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin/intel64:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin:/apps/intel-2020.2/intel-2020.2/debugger_2020/gdb/intel64/bin:/apps/miniconda-4.12.0/bin:/apps/miniconda-4.12.0/condabin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/lib64/qt-3.3/bin:/apps/sbin:/apps/bin:/opt/slurm/bin

**Reply by jprestop** [reply]
@briannen 

Could you please try removing these from your $PATH and then get a clean shell?
```
/apps/miniconda-4.12.0/bin:/apps/miniconda-4.12.0/condabin
```

Thanks!

**Reply by briannen** [reply]
I removed them, then loaded up the met module, and it's still giving me the same error.

Orion-login-4[4] bnelson$ module load contrib
Orion-login-4[5] bnelson$ module load intel/2020.2
Orion-login-4[6] bnelson$ module load met/11.0.1
Orion-login-4[7] bnelson$ /apps/contrib/MET/11.0.1/bin/grid_stat
/apps/contrib/MET/11.0.1/bin/grid_stat: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory
Orion-login-4[8] bnelson$ echo $PATH
/apps/contrib/MET/11.0.1/bin:/apps/contrib/MET/11.0.1/external_libs/bin:/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/bin:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin/intel64:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin:/apps/intel-2020.2/intel-2020.2/debugger_2020/gdb/intel64/bin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/lib64/qt-3.3/bin:/apps/sbin:/apps/bin:/opt/slurm/bin
Orion-login-4[9] bnelson$ module list

Currently Loaded Modules:
  1) contrib/0.1   2) intel/2020.2   3) met/11.0.1

**Reply by jprestop** [reply]
@briannen Thank you.  Could you please run the following and send the output?
```
echo $LD_LIBRARY_PATH
echo $PYTHONPATH
```

**Reply by briannen** [reply]
Orion-login-4[10] bnelson$ echo $LD_LIBRARY_PATH
/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/compiler/lib/intel64_lin:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/ipp/lib/intel64:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/mkl/lib/intel64_lin:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/tbb/lib/intel64/gcc4.8:/apps/intel-2020.2/intel-2020.2/debugger_2020/python/intel64/lib:/apps/intel-2020.2/intel-2020.2/debugger_2020/libipt/intel64/lib:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/daal/lib/intel64_lin:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/daal/../tbb/lib/intel64_lin/gcc4.4:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/daal/../tbb/lib/intel64_lin/gcc4.8:/lib64:/usr/lib64:/lib:/usr/lib:/usr/lib64/qt-3.3/lib:/opt/slurm/lib:/opt/slurm/lib/slurm
Orion-login-4[11] bnelson$ echo $PYTHONPATH

(There is nothing for $PYTHONPATH)

**Reply by jprestop** [reply]
Thank you @briannen. Your outputs are exactly the same as mine.  I modified the permissions of the libpython3.8.so.1.0 file.  Could you please get a clean shell, load the modules and run grid_stat to see if you get the same error?  If you do, please run the following:
```
export LD_LIBRARY_PATH=/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/lib:${PATH}
```
and try again.  Please let me know if one of these end up working.  You shouldn't need to add /work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/lib to your LD_LIBRARY_PATH, but I do expect that to fix the problem you are experiencing.

**Reply by briannen** [reply]
Orion-login-2[1] bnelson$ module load contrib
Orion-login-2[2] bnelson$ module load intel/2020.2
Orion-login-2[3] bnelson$ module load met/11.0.1
Orion-login-2[4] bnelson$ /apps/contrib/MET/11.0.1/bin/grid_stat
**/apps/contrib/MET/11.0.1/bin/grid_stat: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory**
Orion-login-2[5] bnelson$ export LD_LIBRARY_PATH=/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/lib:${PATH}
Orion-login-2[6] bnelson$ module list

Currently Loaded Modules:
  1) contrib/0.1   2) intel/2020.2   3) met/11.0.1

 

Orion-login-2[7] bnelson$ /apps/contrib/MET/11.0.1/bin/grid_stat
**/apps/contrib/MET/11.0.1/bin/grid_stat: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory**
Orion-login-2[8] bnelson$ echo $LD_LIBRARY_PATH
/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/lib:/apps/contrib/MET/11.0.1/bin:/apps/contrib/MET/11.0.1/external_libs/bin:/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/bin:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin/intel64:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin:/apps/intel-2020.2/intel-2020.2/debugger_2020/gdb/intel64/bin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/lib64/qt-3.3/bin:/apps/sbin:/apps/bin:/opt/slurm/bin
Orion-login-2[9] bnelson$ echo $PATH
/apps/contrib/MET/11.0.1/bin:/apps/contrib/MET/11.0.1/external_libs/bin:/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/bin:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin/intel64:/apps/intel-2020.2/intel-2020.2/compilers_and_libraries_2020.2.254/linux/bin:/apps/intel-2020.2/intel-2020.2/debugger_2020/gdb/intel64/bin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/lib64/qt-3.3/bin:/apps/sbin:/apps/bin:/opt/slurm/bin


I'm still getting the same error.

**Reply by jprestop** [reply]
@briannen Ugh.  I see I accidentally typed the incorrect command for you to run.  Could you please get a clean shell and set LD_LIBRARY_PATH in the following way (I accidentally had you add PATH to the end instead of LD_LIBRARY_PATH initially):
```
export LD_LIBRARY_PATH=/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/lib:${LD_LIBRARY_PATH}
```
then load the modules and run grid_stat to see if you get the same error?

**Comment by briannen** [comment]
Yes, same error :(

Orion-login-1[1] bnelson$ export LD_LIBRARY_PATH=/work/noaa/ovp/miniconda/miniconda3/envs/metplus_v5.0_py3.8/lib:${LD_LIBRARY_PATH}
Orion-login-1[2] bnelson$ module list
No modules loaded
Orion-login-1[3] bnelson$ module load contrib
Orion-login-1[4] bnelson$ module load intel/2020.2
Orion-login-1[5] bnelson$ module load met/11.0.1
Orion-login-1[6] bnelson$ module list

Currently Loaded Modules:
  1) contrib/0.1   2) intel/2020.2   3) met/11.0.1

 

Orion-login-1[7] bnelson$ /apps/contrib/MET/11.0.1/bin/grid_stat
/apps/contrib/MET/11.0.1/bin/grid_stat: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

**Reply by jprestop** [reply]
@briannen Thank you.  Unfortunately, I am very surprised and perplexed by this behavior.  Could you please run the following and attach the output file printenv_out.txt here?
```
printenv >& printenv_out.txt
```

**Reply by briannen** [reply]
[printenv_out.txt](https://github.com/dtcenter/METplus/files/11289161/printenv_out.txt)

**Reply by jprestop** [reply]
@briannen Your settings look the same as mine.  Let's plan for a screen sharing session.  I'll set one up and then we can report back here, after we have, hopefully, found the problem and a solution.

**Comment by jprestop** [comment, marked as answer]
A screen sharing session revealed that the directory /work/noaa/ovp/miniconda/ did not have the correct permissions.  Permissions were only open to owner and group, but not other.  Running "chmod -R 755 miniconda" resolved the problem.  Thank you @JohnHalleyGotway and @briannen for meeting and thank you @TaraJensen for suggesting we meet.
